### PR TITLE
Feature - connect & pass traces to the jaeger agent using UDP and thrft.

### PR DESCRIPTION
**Pull request by saktib**
Commit 1:
- Added thrift generator script to pull down jaegertracing/thrift:0.13 
- Ran script to generate Jaeger/Thrift library.
- Modified composer to add "apache/thrift": "^0.13.0" dependency for the Jaeger/Thrift library.

Commit 2:
- Thrift generation script modified, regenerated Thrifts, added Jaeger PSR4 autoloader in the composer to use the namespace Jaeger inside, other classes. 

Commit 3:
- Created Jaeger agent exporter(to handle UDP transports), jaeger agent transporter (wrapper above the thrift udp transporter) & modified the thrift udp transporter to make it a little robust.

Commit 4:
- Make file updated to accommodate 'make thrift'
- .gitignore updated to exclude Jaeger/Thrift
- Removed php8 from Dockerfile and php.yml to support apache/thrift 0.13.0

Commit 5:
- Ran make style, 
- Added span converter for generating Jaeger Thrift Span format array to be used in emitBatch.
- Standardized Jaeger Transport class by implementing the new Transport interface.

Commit 6:
- Fixed sending traces to Jaeger agent using UDP, but it's breaking sending traces to Jaeger collector as in both cases the traceId and spanId requires to be different datatype. Notes below....

Note:
Jaeger agent (using Jaeger UDP)accepts traceId and spanId in I64 format (64bit Integer) from the client which internally converts it to 32 char hex string as traceId and 16char hex string as spanId and send it to the Jaeger collector.

Jaeger collector (using Zipkin http) accepts 32 char hex string as traceId and 16char hex string as spanId from the client. 

Need a fix to differentiate the data type and validations around it.

### CODE REVIEW

File: script/thrift_gen.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#!/bin/sh
```
Comment (bobstrecansky) [inline script/thrift_gen.sh:1]: We should also add an entry for this in the [Makefile](https://github.com/open-telemetry/opentelemetry-php/blob/main/Makefile#L27)

Comment (bobstrecansky) [reply script/thrift_gen.sh:1]: This file also needs executable permissions added to it so we can generate thrift output.  `chmod +x script/thrift_gen.sh` should handle this for you.

File: script/thrift_gen.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+cd "$(dirname "$0")/.."
+
+# checkout jaeger thrift files
+rm -rf jaeger-idl
+git clone https://github.com/jaegertracing/jaeger-idl
+
+# define thrift cmd
+THRIFT="docker run -u $(id -u) -v '${PWD}:/data' jaegertracing/thrift:0.13 thrift -o /data/jaeger-idl"
+THRIFT_CMD="${THRIFT} --gen php:psr4,oop"
```
Comment (bobstrecansky) [inline script/thrift_gen.sh:13]: The output from when I ran this locally was:

```
docker run -u 502 -v '/Users/bstrecansky/git/opentelemetry-php:/data' jaegertracing/thrift:0.13 thrift -o /data/jaeger-idl --gen php:psr4,oop /data/jaeger-idl/thrift/baggage.thrift
[WARNING:generation:1] psr4 is default option! needn't add psr4 option!
```

Perhaps you don't need to explicitly state the psr4 option?  Were you adding this for succinctness?

File: Jaeger/ThriftUdpTransport.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
+<?php
+
+namespace Jaeger;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
```
Comment (bobstrecansky) [inline Jaeger/ThriftUdpTransport.php:6]: TIL about `use Psr\Log\NullLogger;` - very slick!

**Review by bobstrecansky — COMMENTED** [review]
This is a great start!  A couple things I see that would be helpful to add:

* You may want to add the thrift generated files to the .gitignore so they aren't accidentally committed.
* Perhaps we could use a directory structure of `./thrift/Jaeger` rather than `./Jaeger/thrift` - that may make things a bit more clear for others looking for the thrift protos
* It'd be nice to have some test cases or examples for this implementation.
* A couple other comments in line.

File: script/thrift_gen.sh (hunk 1 above)
Comment (saktib) [reply script/thrift_gen.sh:1]: - executable permission given to script/thrift_gen.sh
- Makefile execution failing due to [Dockerfile](https://github.com/open-telemetry/opentelemetry-php/blob/127166bc29f9fb7bf96ab5250a6118b7585d5e2f/docker/Dockerfile#L1) PHP:8 version as [apache/thrift](https://packagist.org/packages/apache/thrift) latest verion is compatible with php: ^5.5 || ^7.0 
- I modified  Dockerfile locally to use php:7.3-buster and it worked fine.

Let me know your thought on this, as  apache/thrift is not supporting PHP:8

File: script/thrift_gen.sh (hunk 2 above)
Comment (saktib) [reply script/thrift_gen.sh]: Removed psr4 explicit check as it's already the default option.

File: script/thrift_gen.sh (hunk 1 above)
Comment (bobstrecansky) [reply script/thrift_gen.sh:1]: We talked about this in the SIG meeting today.  We are going to remove the hard dependency on PHP8 for now.  There is an apache thrift issue open about this:

https://github.com/apache/thrift/pull/2288

Once this issue has been closed, we can re-enable PHP8 as a dependency.

You can feel free to change this dependency for the time being:

* Change the dockerfile to use php:7.4-buster: https://github.com/open-telemetry/opentelemetry-php/blob/main/docker/Dockerfile#L1
* remove php 8.0 from the github action workflow : https://github.com/open-telemetry/opentelemetry-php/blob/main/.github/workflows/php.yml#L17

Comment (bobstrecansky) [reply script/thrift_gen.sh:1]: @saktib - once you complete this, I think we'll be able to merge this?

Comment (saktib) [reply script/thrift_gen.sh:1]: Yes, absolutely. I'll commit the above fixes soon.

**Comment by bobstrecansky** [comment]
You'll need to run your code through a `make style` and commit it to get rid of that linting error.

**Comment by bobstrecansky** [comment]
@saktib - besides the `make style` to lint your code, is there anything else you need for this PR?

**Comment by beniamin** [comment]
Nice work here.
I am think that it may fit better in the `OpenTelemetry\Contrib` namespace. What do you say?

**Comment by saktib** [comment]
> @saktib - besides the `make style` to lint your code, is there anything else you need for this PR?

@bobstrecansky , almost everything is fixed apart from one where trace-id and span-ids are going 0s from agent to collector even the library is passing it correctly. Working on it.... Ran make style. phan is throwing up with a couple of items will fix those as well.

**Comment by bobstrecansky** [comment]
@saktib - are you still planning on fixing these things?  Would love to help you get this PR across the finish line!

**Comment by saktib** [comment]
> @saktib - are you still planning on fixing these things? Would love to help you get this PR across the finish line!

@bobstrecansky , yes I'm looking forward to your input on datatype for trace & span ids. As we discussed UDP thrift to jaeger agent supports 64-bit integer and HTTP json to jaeger collector needs hex string. 

This PR is having a working flow for  UDP thrift to jaeger agent that supports 64-bit integer, however breaking the HTTP way.
If you could confirm we can have the datatype as I64 for trace and span through the library and while doing the http call will convert the I64 to hex and pass it to the collector. 

You can answer it here, or I can able to attend the next SIG meet to finalize the flow.

**Comment by bobstrecansky** [comment]
@saktib - have you looked to see what other language sigs are doing for this particular use case?  Changing this to an I64 would break spec I believe, which we are actively trying not to do.

**Comment by bobstrecansky** [comment]
Just following up here.

**Comment by saktib** [comment]
> Just following up here.

I'll look into what other language sigs are doing for this case, got busy with the integration of this updated client which now sends traces to jaeger agent and the outcome comes out great as the http calls to the collector cluster have reduced. 

image

**Comment by bobstrecansky** [comment]
Glad to see this is better for y'all - @lalex fixed a [memory leak](https://github.com/open-telemetry/opentelemetry-php/pull/286) that was lingering; I'm assuming that had something to do with this working better for your implementation.

**Comment by saktib** [comment]
> @saktib - have you looked to see what other language sigs are doing for this particular use case? Changing this to an I64 would break spec I believe, which we are actively trying not to do.

Hi Bob,As  I've looked opentelemetry client library written in Java and Go and all seem to be using int64 when it's coming to transport via UDP to Jaeger agent using thrift, as thrift library itself has a validation to get traceid and spanid in int64 format. And while it's http transport they are using hex string.

So the solution we need to look out for is 

- Keep int64 while transporting data to jaeger agent via UDP (which is happening currently on this PR) and modify the PR to support hex string for spanid, traceid while transporting data to jaeger collector via http.

Let me know your thought.

**Comment by bobstrecansky** [comment]
That sounds like a good solution to me.  Thank you for verifying your strategy with other SIG work!

**Comment by saktib** [comment]
> That sounds like a good solution to me. Thank you for verifying your strategy with other SIG work!

great! will proceed with this solution.

**Comment by bobstrecansky** [comment]
@saktib -  are you still planning on pursuing the rest of this PR?

**Comment by saktib** [comment]
> @saktib - are you still planning on pursuing the rest of this PR?

Hi @bobstrecansky, really sorry for the delay on this, caught up with a few other items. Will try to close this over this weekend.

**Comment by saktib** [comment]
Hi @bobstrecansky , Just to give you updates on the new commits. I found this PR from master branch of my fork is quite outdated and I see the mater branch itself is moved to main and I found a lot of conflicts while rebasing. 

So I created a separate branch in my fork from the latest main branch and copied over all the changes from here to there and added changes to support UDP & HTTP... currently doing a few rounds of testing before raising a PR from the newly created branch. Once that's approved we can close this PR and merge the newly created PR to main. 

Hopefully by coming Friday EOD i should be able to put out the new PR.

**Comment by bobstrecansky** [comment]
Closed in favor of : https://github.com/open-telemetry/opentelemetry-php/pull/337

**Pull request closed without merging**